pennylane/operation.py

```diff
@@ -2544,6 +2544,29 @@ def heisenberg_obs(self, wire_order):
         return self.heisenberg_expand(U, wire_order)
 
 
+class StatePrep(Operation):
+    """An interface for state-prep operations."""
+
+    grad_method = None
+    _queue_category = "_prep"
+
+    # pylint:disable=too-few-public-methods
+    @abc.abstractmethod
+    def state_vector(self, wire_order=None, interface="autograd"):
```

Comment:
Or should we make `"numpy"` the default interface?

Reply:
we could, but I'm just copying [what qnode does](https://github.com/PennyLaneAI/pennylane/blob/c05315f93a32c5220cb6bf0ac2aa0456d6a71fa8/pennylane/qnode.py#L352)

Reply:
I'm probably missing a lot of context, but how come there is a default here?

Reply:
we could feasibly do without it, but the intention of this PR is to prepare for the addition of a `create_initial_state` function (like [here](https://github.com/PennyLaneAI/pennylane/pull/3654/commits/e024697d64b509776aebee08825da475a9573215#diff-eeb139fd2df15a47f705d5efc436f514bfb47d15afc7653fa680f01aca946cf9R755-R776), but I'm doing it in another PR after this one). That function will be used at the beginning of our new `qml.execute()`, calling a state-prep operator (from this PR) if one is present in the tape. Other things we could do:
- make it a requirement (positional arg). I could see users calling `op.state_vector()` and not want to provide an interface. autograd is the default elsewhere, so keeping this would be consistent, and users wouldn't have to explicitly declare the default behaviour they want
- infer the interface from `self.parameters`. I wouldn't do this because some implementations might not expect this. I even think `BasisState` could receive a vanilla Python list most of the time, no ML framework is needed to specify an array of 0s and 1s
- default to `numpy`. It's a change from QNode, but I wouldn't mind it. We should communicate and agree on this change if it's what we want to do.

Reply:
🤔 What is the intended logical flow here? Are we expecting devices to programmatically call `op.state_vector()` as part of the execution pipeline?

In which case, the state prep operation might accept a _trainable_ parameter -- so I would suggest option 2 -- infer the interface from the parameter.

Regarding having a default - I am hesitant, partly because this is something we want to start removing or changing at the QNode level 🤔 @trbromley

Reply:
I'm all good with inferring the interface!

Reply:
@josh146 Just to clarify since there was some miscommunication because of how many things we have called `execute`.  This will be the first step inside `PythonDevice.execute` when we are performing a simulation using the python device.

It will be used by the next-gen `default.qubit` instead of `DefaultQubit._apply_basis_state` and `DefaultQubit._apply_state_vector`.